org.eclipse.xtext/src/org/eclipse/xtext/validation/LocalUniqueNameContext.java

```diff
@@ -0,0 +1,189 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Sebastian Zarnekow and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.xtext.validation;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.eclipse.emf.ecore.EClass;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.EcorePackage;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.naming.QualifiedName;
+import org.eclipse.xtext.resource.EObjectDescription;
+import org.eclipse.xtext.resource.IEObjectDescription;
+import org.eclipse.xtext.resource.ISelectable;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.validation.INamesAreUniqueValidationHelper.Context;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.FluentIterable;
+
+/**
+ * <p>A case-sensitive, local validation context to check unique names in in the contents of 
+ * a container {@link EObject} or within a list.
+ * </p>
+ * Sample usage in a validation:
+ * 
+ * <pre>
+ * class MyDslValidator extends AbstractDeclarativeValidator {
+ * 	&#64;Inject
+ * 	INamesAreUniqueValidationHelper helper;
+ * 
+ * 	&#64;Check
+ * 	public void checkUniqueNames(Model model) {
+ * 		helper.checkUniqueNames(new LocalUniqueNameContext(model), this);
+ * 	}
+ * }
+ * </pre>
+ * 
+ * @see INamesAreUniqueValidationHelper#checkUniqueNames(Context, ValidationMessageAcceptor)
+ * 
+ * @since 2.22
+ * @author Sebastian Zarnekow - Initial contribution and API
+ */
+public class LocalUniqueNameContext implements INamesAreUniqueValidationHelper.Context, ISelectable {
+
+	private final List<IEObjectDescription> objectsToValidate;
+	private final Map<String, Object> index;
+	private final CancelIndicator ci;
+
+	public LocalUniqueNameContext(EObject container, Function<EObject, String> nameFunction, CancelIndicator ci) {
```

Comment:
any reason why we use String and not QualifiedName?

Reply:
Locally its likely more convenient to work with the name directly. Do you think it would be added value to work with QN here?

Reply:
no. i just stumbled over it